nrf-hal-common/src/bbq_uarte/irq.rs

```diff
@@ -0,0 +1,179 @@
+    }
+
+    pub fn interrupt(&mut self) {
+        let endrx = self.uarte.events_endrx.read().bits() != 0;
+        let endtx = self.uarte.events_endtx.read().bits() != 0;
+        let rxdrdy = self.uarte.events_rxdrdy.read().bits() != 0;
+        let error = self.uarte.events_error.read().bits() != 0;
+        let txstopped = self.uarte.events_txstopped.read().bits() != 0;
+
+        let timeout = self.timeout_flag.swap(false, SeqCst);
+        let errsrc = self.uarte.errorsrc.read().bits();
+
+        // RX section
+        if endrx || timeout || self.rx_grant.is_none() {
+            // We only flush the connection if:
+            //
+            // * We didn't get a "natural" end of reception (full buffer), AND
+            // * The timer expired, AND
+            // * We have received one or more bytes to the receive buffer
+            if !endrx && timeout && rxdrdy {
+                uarte_cancel_read(&self.uarte);
+            }
+
+            compiler_fence(SeqCst);
+
+            // Get the bytes received. If the rxdrdy flag wasn't set, then we haven't
+            // actually received any bytes, and we can't trust the `amount` field
+            // (it may have a stale value from the last reception)
+            let amt = if rxdrdy {
+                self.uarte.rxd.amount.read().bits() as usize
+            } else {
+                0
+            };
+
+            // If we received data, cycle the grant and get a new one
+            if amt != 0 || self.rx_grant.is_none() {
+                let gr = self.rx_grant.take();
+
+                // If the buffer was full last time, we may not actually have a grant right now
+                if let Some(gr) = gr {
+                    gr.commit(amt);
+                }
+
+                // Attempt to get the next grant. If we don't get one now, no worries,
+                // we'll try again on the next timeout
+                if let Ok(mut gr) = self.incoming_prod.grant_exact(self.block_size) {
+                    uarte_start_read(&self.uarte, &mut gr).unwrap();
+                    self.rx_grant = Some(gr);
+                }
+            }
+        }
+
+        // TX Section
+        if endtx || self.tx_grant.is_none() {
+            if endtx {
+                if let Some(gr) = self.tx_grant.take() {
+                    let len = gr.len();
+                    gr.release(len.min(EASY_DMA_SIZE));
+                }
+            }
+
+            if let Ok(gr) = self.outgoing_cons.read() {
+                let len = gr.len();
+                uarte_start_write(&self.uarte, &gr[..len.min(EASY_DMA_SIZE)]).unwrap();
+                self.tx_grant = Some(gr);
+            }
+        }
+
+        // Clear events we processed
+        if endrx {
+            self.uarte.events_endrx.write(|w| w);
+        }
+        if endtx {
+            self.uarte.events_endtx.write(|w| w);
+        }
+        if error {
+            self.uarte.events_error.write(|w| w);
+        }
+        if rxdrdy {
```

Comment:
The RX FIFO is only flushed if !endrx, but the rxrdy will still be cleared if set.
If e.g. more than block_size but less than 2*block_size is received, rxdrdy will be cleared and therefore will not be flushed on the following timeout